pkg/kubelet/cm/cpuset/cpuset.go

```diff
@@ -0,0 +1,260 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cpuset
+
+import (
+	"bytes"
+	"fmt"
+	"github.com/golang/glog"
+	"reflect"
+	"sort"
+	"strconv"
+	"strings"
+)
+
+// CPUSet is a set-like data structure for CPU IDs.
+type CPUSet map[int]struct{}
+
+// NewCPUSet return CPUSet based on provided cpu id's
+func NewCPUSet(cpus ...int) CPUSet {
+	res := CPUSet{}
+	for _, c := range cpus {
+		res.Add(c)
+	}
+	return res
+}
+
+// Size returns the number of elements in this set.
+func (s CPUSet) Size() int {
+	return len(s)
+}
+
+// IsEmpty returns true if there are zero elements in this set.
+func (s CPUSet) IsEmpty() bool {
+	return s.Size() == 0
+}
+
+// Contains returns true if the supplied element is present in this set.
+func (s CPUSet) Contains(cpu int) bool {
+	_, found := s[cpu]
+	return found
+}
+
+// Add mutates this set to contain the supplied elements.
+func (s CPUSet) Add(cpus ...int) {
```

Comment:
Yes there is. Add and Remove are not safe to call in a concurrent context. Is it sufficient to document this? The expected use of Add and Remove follows the examples in this package, mutating "owned" references as in:

```go
func myCoolFunction() cpuset.CPUSet {
    result := cpuset.NewCPUSet()
    for i := 0; i < 10; i++ {
        result.Add(foo)
    }
    return result
}
```

If a doc update isn't sufficient, I would prefer not to add an RWLock to this type. Another option could be to encode the expected use by creating a mutable builder type from which the user can retrieve the immutable result CPUSet (I quite like this idea). That could look something like this:

```go
func myCoolFunction() cpuset.CPUSet {
    builder := cpuset.NewCPUSetBuilder()
    for i := 0; i < 10; i++ {
        builder.Add(foo)
    }
    return builder.Result()
}
```